File: src/leetCode/c++/Solution53.cpp
//Greedy Kadane's Algorithm O(n) O(1)
class Solution53 {
public:
    int maxSubArray(vector<int>& nums) {
        int res = INT_MIN;
        int sum = 0;
        for(int num : nums){
            sum = max(sum + num, num);//若num較大, 則捨去之前
            res = max(res, sum);
        }
        return res;
    }
};

File: java/Solution53.java
package leetCode.java;

//Greedy Kadane's Algorithm O(n) O(1)
class Solution53 {
    public int maxSubArray(int[] nums) {
        int sum = 0, res = Integer.MIN_VALUE;;
        for (int num : nums) {
            sum = Math.max(sum + num, num); // 若num比較大, 則捨去之前加總的數列
            res = Math.max(res, sum);
        }
        return res;
    }
}
/*
sum = 以目前位置結尾的最大 subarray sum
res = 到目前為止最大的答案
num本體更大的話, 直接拋棄之前的

nums=   -2, 1,      -3, 4,      -1, 2,  1,  -5, 4
sum=    -2  -1->1   -2  2->4    3   5   6   1   5


*/